File: src/main/java/com/github/aivancioglo/resttest/http/Endpoint.java
package com.github.aivancioglo.resttest.http;

import java.net.MalformedURLException;
import java.net.URL;

import static io.restassured.http.Method.*;

/**
 * Simple endpoint for simple request.
 */
public class Endpoint extends HTTPRequest<Endpoint> {
    private static final String PROTOCOL = "com/github/aivancioglo/resttest/http";

    public Endpoint() {
        setProtocol(PROTOCOL);
    }

    /**
     * Method for GET request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse get() {
        return send(GET);
    }

    /**
     * Method for POST request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse post() {
        return send(POST);
    }

    /**
     * Method for PATCH request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse patch() {
        return send(PATCH);
    }

    /**
     * Method for DELETE request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse delete() {
        return send(DELETE);
    }

    /**
     * Method for PUT request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse put() {
        return send(PUT);
    }

    /**
     * Method for HEAD request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse head() {
        return send(HEAD);
    }

    /**
     * Method for TRACE request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse trace() {
        return send(TRACE);
    }

    /**
     * Method for OPTIONS request.
     *
     * @return HTTPResponse instance.
     */
    public HTTPResponse options() {
        return send(OPTIONS);
    }

    /**
     * Set URL for request.
     *
     * @param endpoint String.
     * @return instance of this class.
     */
    public Endpoint setURL(String endpoint) {
        URL url;

        try {
            url = new URL(endpoint);
        } catch (MalformedURLException e) {
            throw new RuntimeException(e);
        }

        setProtocol(url.getProtocol());
        setHost(url.getHost());
        setPath(url.getPath() + "?" + url.getQuery());

        return this;
    }
}

File: src/main/kotlin/com/github/aivancioglo/resttest/http/Endpoint.kt
package com.github.aivancioglo.resttest.http

import com.github.aivancioglo.resttest.setters.Setter
import com.github.aivancioglo.resttest.setters.Setters.Companion.host
import com.github.aivancioglo.resttest.setters.Setters.Companion.path
import com.github.aivancioglo.resttest.setters.Setters.Companion.port
import com.github.aivancioglo.resttest.setters.Setters.Companion.protocol
import io.restassured.http.Method.*
import java.net.URL

/**
 * Universal class for any HTTP requestSpecification making.
 */
class Endpoint : Request() {
    /**
     * Making GET requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun get(vararg setters: Setter) = send(GET, *setters)

    /**
     * Making GET requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun get(url: String, vararg setters: Setter): Response {
        set(url)
        return send(GET, *setters)
    }

    /**
     * Making POST requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun post(vararg setters: Setter) = send(POST, *setters)

    /**
     * Making POST requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun post(url: String, vararg setters: Setter): Response {
        set(url)
        return send(POST, *setters)
    }

    /**
     * Making PUT requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun put(vararg setters: Setter) = send(PUT, *setters)

    /**
     * Making PUT requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun put(url: String, vararg setters: Setter): Response {
        set(url)
        return send(PUT, *setters)
    }

    /**
     * Making PATCH requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun patch(vararg setters: Setter) = send(PATCH, *setters)

    /**
     * Making PATCH requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun patch(url: String, vararg setters: Setter): Response {
        set(url)
        return send(PATCH, *setters)
    }

    /**
     * Making DELETE requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun delete(vararg setters: Setter) = send(DELETE, *setters)

    /**
     * Making DELETE requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun delete(url: String, vararg setters: Setter): Response {
        set(url)
        return send(DELETE, *setters)
    }

    /**
     * Making OPTIONS requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun options(vararg setters: Setter) = send(OPTIONS, *setters)

    /**
     * Making OPTIONS requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun options(url: String, vararg setters: Setter): Response {
        set(url)
        return send(OPTIONS, *setters)
    }

    /**
     * Making HEAD requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun head(vararg setters: Setter) = send(HEAD, *setters)

    /**
     * Making HEAD requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun head(url: String, vararg setters: Setter): Response {
        set(url)
        return send(HEAD, *setters)
    }

    /**
     * Making TRACE requestSpecification.
     *
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun trace(vararg setters: Setter) = send(TRACE, *setters)

    /**
     * Making TRACE requestSpecification.
     *
     * @param url of your requestSpecification.
     * @param setters are setting up requestSpecification specification.
     * @return Response instance.
     */
    fun trace(url: String, vararg setters: Setter): Response {
        set(url)
        return send(TRACE, *setters)
    }

    /**
     * Setting url to URL to requestSpecification specification.
     *
     * @param endpoint of requestSpecification.
     */
    private fun set(endpoint: String) {
        var link = endpoint

        if (!endpoint.matches(Regex("^\\w+?://.*?")))
            link = "http://$link"

        val url = URL(link)

        if (url.port != -1)
            set(port(url.port))

        set(
                protocol(url.protocol),
                host(url.host),
                path("${url.path}?${url.query}")
        )
    }
}